### PR TITLE
Add is_mathml_annotation_xml_integration_point method

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
r? @nox

#119 

[Review on Reviewable](https://reviewable.io/reviews/servo/html5ever/186)

### CODE REVIEW

File: src/rcdom.rs [hunk 1]
```diff
@@ -195,6 +196,14 @@ impl TreeSink for RcDom {
     }
 
     fn create_element(&mut self, name: QualName, attrs: Vec<Attribute>) -> Handle {
+        for it in 0..attrs.len() {
+            match attrs[it].name {
+                qualname!(html, "encoding") => {
+                    return new_node(Element(name, AnnotationXml(attrs[it].value.to_lowercase() == "text/html"), attrs));
```
Comment (SimonSapin) [inline src/rcdom.rs]: Use `.eq_ignore_ascii_case("text/html")` to save a memory allocation.

File: src/rcdom.rs [hunk 2]
```diff
@@ -45,6 +45,7 @@ pub enum ElementEnum {
     /// A template element and its template contents.
     /// https://html.spec.whatwg.org/multipage/#template-contents
     Template(Handle),
+    AnnotationXml(bool),
```
Comment (SimonSapin) [inline src/rcdom.rs]: What does it mean for an element to "be" this variant? Is it exclusive with other variants? What does the boolean mean? Consider using a struct-like variant in order to name the boolean. Please add a doc-comment with a link to the relevant section of spec.

File: src/tree_builder/interface.rs [hunk 3]
```diff
@@ -133,6 +133,10 @@ pub trait TreeSink {
     fn complete_script(&mut self, _node: Self::Handle) -> NextParserState {
         NextParserState::Continue
     }
+
+    // Returns true if the adjusted current node is an HTML integration point
+    // and the token is a start tag
+    fn is_mathml_annotation_xml_integration_point(&self) -> bool;
```
Comment (nox) [inline src/tree_builder/interface.rs]: This should take a `Handle` and that's the thing on which you check whether it is an integration point or not.

File: src/rcdom.rs [hunk 4]
```diff
@@ -303,6 +312,13 @@ impl TreeSink for RcDom {
             panic!("not a script element!");
         }
     }
+
+    fn is_mathml_annotation_xml_integration_point(&self) -> bool {
+        match (**self.document).borrow().node {
```
Comment (nox) [inline src/rcdom.rs]: The thing you need to check is the given `Handle` that needs to be passed to the method.

File: src/rcdom.rs (hunk 2 above)
Comment (GuillaumeGomez) [inline src/rcdom.rs]: I didn't add comments yet, it was just supposed to be a first shot to get help. on some details before finishing it. I should have precised, my bad.

**Comment by GuillaumeGomez** [comment]
Ok, I updated the current code @nox.

File: src/rcdom.rs [hunk 5]
```diff
@@ -303,6 +313,13 @@ impl TreeSink for RcDom {
             panic!("not a script element!");
         }
     }
+
+    fn is_mathml_annotation_xml_integration_point(&self, handle: &Self::Handle) -> bool {
+        match (**handle).borrow().node {
+            Element(_, AnnotationXml(ret), _) => ret,
+            _ => false,
```
Comment (nox) [inline src/rcdom.rs]: This should be unreachable.

File: src/rcdom.rs [hunk 6]
```diff
@@ -195,6 +197,14 @@ impl TreeSink for RcDom {
...
 
     fn create_element(&mut self, name: QualName, attrs: Vec<Attribute>) -> Handle {
+        for it in 0..attrs.len() {
+            match attrs[it].name {
+                qualname!(html, "encoding") => {
+                    return new_node(Element(name, AnnotationXml(attrs[it].value.eq_ignore_ascii_case("text/html")), attrs));
+                }
+                _ => continue,
+            }
+        }
```
Comment (nox) [inline src/rcdom.rs]: This should be in the match expression defining `info` below, for `qualname!(mathml, "annotation-xml")` elements.

**Comment by GuillaumeGomez** [comment]
Anything else to add/update?

File: src/tree_builder/actions.rs [hunk 7]
```diff
@@ -901,11 +901,15 @@ impl<Handle, Sink> TreeBuilderActions<Handle>
 
         if let qualname!(mathml, "annotation-xml") = name {
             if let TagToken(Tag { kind: StartTag, name: atom!("svg"), .. }) = *token {
```
Comment (nox) [inline src/tree_builder/actions.rs]: This condition should go away. It has nothing to do here.

File: src/tree_builder/interface.rs [hunk 8]
```diff
@@ -133,6 +133,12 @@ pub trait TreeSink {
     fn complete_script(&mut self, _node: Self::Handle) -> NextParserState {
         NextParserState::Continue
     }
+
+    // Returns true if the adjusted current node is an HTML integration point
+    // and the token is a start tag
+    fn is_mathml_annotation_xml_integration_point(&self, handle: &Self::Handle) -> bool {
```
Comment (nox) [inline src/tree_builder/interface.rs]: This should take `Self::Handle`, and you should clone handles when passing them.

File: src/tree_builder/actions.rs (hunk 7 above)
Comment (nox) [inline src/tree_builder/actions.rs]: I was wrong about this.

**Pull request merged**